### ISSUE TITLE
nvim_buf_get_lines() only works when buffer has a window open

### ISSUE THREAD

**Issue by phodge** [neovim/neovim#5534]
- `nvim --version`: NVIM v0.1.5
- Operating system/version: OS X 10.11.6
- Terminal name/version: iTerm2 3.0.10
- `$TERM`: `"screen-256-color"`

The neovim API function `nvim_buf_get_lines()` will return a list with a single empty string if the nominated buffer isn't currently open in any windows.
### Steps to reproduce using python client

```
#/usr/bin/env python3
file1 = '/tmp/test.txt'
with open(file1, 'w') as f: f.write("A\nB\nC\n")
from neovim import attach
nvim = attach('child', argv=['nvim', '--embed', '-u', 'NONE', file1])
print('nvim.buffers[1][:] = %r' % nvim.buffers[1][:])
nvim.command('enew')
print('nvim.buffers[1][:] = %r' % nvim.buffers[1][:])
nvim.quit()
```
### Expected output

```
nvim.buffers[1][:] = ['A', 'B', 'C']
nvim.buffers[1][:] = ['A', 'B', 'C']
```
### Actual output

```
nvim.buffers[1][:] = ['A', 'B', 'C']
nvim.buffers[1][:] = ['']
```
### Further notes

The behaviour might not have to do with whether or not the buffer is visible in any windows, but that is the easiest way for me to reproduce it right now. It's because the buffer's `buf->b_ml` memline is empty, which seems to happen when the buffer is no longer visible in any windows.

**Comment by bfredl** [comment]
Do you have `set hidden?`

**Comment by justinmk** [comment]
Same question as @bfredl. And the next question is, what should `nvim_buf_get_lines()` do in that case? Load the buffer or return an error?

**Comment by bfredl** [comment]
Currently we actually replicate vim behavior, if one has `nohidden` and does
`:py3 b = vim.current.buffer`

saves and loads another buffer (so the first buffer is not visible anymore)

`:py3 print (b[:])` 

gives `['']` in both nvim and vim. Purely from an _api_ point of view (setting the python compatibilIty question aside) I would prefer if this was an error. We should at least allow the api to check if the buffer is loaded (maybe already is through some option?). If the user expects to /use a plugin that needs to/ background process buffers, they should `set hidden`.

**Comment by justinmk** [comment]
> If the user expects to /use a plugin that needs to/ background process buffers, they should set hidden.

For plugin's buffers, they can also `:setlocal bufhidden=hide` to avoid being affected by the global `'hidden'` option. 

But I strongly agree that there should at least be an error at the API level.

**Comment by phodge** [comment]
I don't have `hidden` turned on. Also I wasn't aware that this behaviour has actually come across from vim.

If an API error is added as the fix, there is still the problem that the standard practise for getting vim buffer contents when you don't know if the buffer is visible must be:
```
# use vertical split to reduce the likelyhood of vim not having enough room to open a new window
nvim.command('vert sbuf%d' % bufnr)
lines = nvim.buffers[bufnr][:]
# close the window you just opened
nvim.command('close')
```

Jumping through hoops like this to avoid unexpected errors is a large part of what makes writing vim plugins such a nightmare, and if we are going to advise plugin authors to explicitly open the buffer in a new window anyway, maybe the API should just load and unload the buffer automatically and make plugin writing a little bit easier. /my 2c

**Comment by phodge** [comment]
This will be resolved by https://github.com/neovim/neovim/pull/8660

**Comment by justinmk** [comment]
closed in #8660
